### ISSUE TITLE
Seems to be #265ing

### ISSUE THREAD

**Issue by oxinabox** [FluxML/Zygote.jl#22]
Maybe there is a catch that is not getting updated when a function is redefined?

# Code:
```
using Zygote;
g(x) = 2x;
@show g'(10);
g(x) = 3x;
@show g'(10);
```
Expected output is **2**, then **3**
Actual output is **2**, then **2** again.

## Demo:

```
julia> using Zygote;

julia> g(x) = 2x;

julia> @show g'(10);
(g')(10) = 2

julia> g(x) = 3x;

julia> @show g'(10);
(g')(10) = 2
```

**Comment by MikeInnes** [comment]
Yes, this is mentioned in the readme; you can call `Zygote.refresh()` to avoid it.

**Comment by oxinabox** [comment]
Ah, I missed that.
Worth leaving this open, to track against it?

I feel like hooking into the julia #265,
should be doable?

**Comment by oxinabox** [comment]
thoughts, on the code I posted before?
I think it can be made sufficiently smart to solve this without a large cost overhead

**Comment by MikeInnes** [comment]
So the idea is that we call `refresh()` at the entry point of `gradient`? And then that can call `eval`/`invokelatest` on `_forward` to get the latest definition.

It seems hard to do this in general. I could see it working in the case that everything is fully well-typed, but what about if the functions that `f` calls are not known at compile time? We either make that significantly more expensive (call `refresh` again at the boundary) or just push this issue into more complex code (which might actually make it more surprising when it comes up). I could see something like this working, but it would be a relatively invasive change.

**Comment by oxinabox** [comment]
> So the idea is that we call refresh() at the entry point of gradient? 

Yes, we call a version of `refresh`  before the existing gradient, where this version only refreshes if required according to `world_age`s.

And calculating world ages has to be done recursively, since only functions directly changed have the `min_world` field updated  (I am pretty sure you understand that already, but just for anyone else reading this in the future. Like future-me.)

> I could see it working in the case that everything is fully well-typed, but what about if the functions that f calls are not known at compile time? 

I assume you mean _if the **methods** called are not known at compile time.
Since the functions always are (excluding maybe some kinda `eval` hackery that I'm not sure is actually possible.).

Yes, if the method called is not known at compile time, the world-age needs to be (recursively) checked for all possibilities.  And that set can be lowered somewhat by various things like argument counting (doesn't work for splatting), and knowing some of the argument types, even if not all of them.

The code in https://github.com/FluxML/Zygote.jl/issues/22#issuecomment-425779953
always checks all methods, it does none of the cutting down of things just mentioned.
and so is quiet slow.


A heuristic to make this much faster would be a max nesting depth for how deep to check for changes in called methods.
Only checking for modifications in the function being calls world age is really fast.
(No recursion, as in `min_world(f) = maximum(mm.min_world for mm in methods(f).ms)`)

**Comment by MikeInnes** [comment]
> I assume you mean _if the methods called are not known at compile time. Since the functions always are ...

Slightly contrived, but for example:

```julia
function foo(fs)
  f = pop!(fs
  f(1, 2)
end

foo(Any[+])
```

**Comment by willow-ahrens** [comment]
Would it make sense to call the generators with `Core._apply_pure()` as is done in https://github.com/NHDaly/StagedFunctions.jl/blob/master/src/StagedFunctions.jl#L116? You don't need to invalidate later like StagedFunctions.jl does, but it would at least free things up so that if you define relevant rules before calling the generator you wouldn't need to `refresh()`. This is the approach I'm taking with https://github.com/willow-ahrens/Finch.jl/pull/176 and it seems to work okay for me so far.

**Comment by ToucheSir** [comment]
There are some changes to world age handling being discussed in https://github.com/FluxML/IRTools.jl/pull/109, but I lack the know-how to say how they fit into your proposal.

**Comment by willow-ahrens** [comment]
Does this mean that zygote no longer has 265 issues? #1560 doesn't seem to modify code that would affect this?

**Comment by pxl-th** [comment]
> Does this mean that zygote no longer has 265 issues? [#1560](https://github.com/FluxML/Zygote.jl/pull/1560) doesn't seem to modify code that would affect this?

That PR doesn't do anything regarding this, but I've checked that MWE above no longer reproduces the issue

**Comment by willow-ahrens** [comment]
If nobody intended to fix this, I highly doubt the underlying issue is fixed, though it could be the above example does not reproduce. Was a change made to base Julia regarding 265?

**Comment by willow-ahrens** [comment]
It would be great to know which changes were made. Can we reproduce this error on LTS 10.7?
